File: cpp/tensorrt_llm/manifold/worker.cpp
#include <thread>
#include <vector>
#include <functional>
#include <iostream>
#include <cuda_runtime.h>
#include <c10/cuda/CUDAStream.h>

#include "worker.h"

namespace manifold{
// Class Worker
Worker::Worker(int tid, int gpu_id, const std::function<void()>& f): tid_(tid), gpu_id_(gpu_id), recv_buf_ptr_(nullptr), recv_buf_size_(0), send_end_(false) {  
    thd_ = std::make_unique<std::thread>([this, f]() {
        auto ident_ = std::hash<std::thread::id>{}(std::this_thread::get_id());
        Controller::GetInstance()->add_idmap(ident_, tid_);
        f();
    });
}

int Worker::get_tid() {
    return tid_;
}

int Worker::get_gpu_id() {
    return gpu_id_;
}

void Worker::recv_async(void* recv_buf, size_t recv_size)
{
    {
        std::lock_guard<std::mutex> lg(mtx__);
        recv_buf_queue_.push({recv_buf, recv_size});  
    }
    cv__.notify_all();
}

void Worker::send_async(int peer, const void* src, size_t src_size, cudaStream_t stream) {
    auto peer_worker = Controller::GetWorker(peer);
    std::pair<void*, size_t> recv_buf_pair;

    {
        std::unique_lock<std::mutex> ul(peer_worker->mtx__);
        peer_worker->cv__.wait(ul, [peer_worker] { return !peer_worker->recv_buf_queue_.empty();});
        recv_buf_pair = peer_worker->recv_buf_queue_.front();
        peer_worker->recv_buf_queue_.pop();
    }
    
    void* recv_buf_ = recv_buf_pair.first;
    size_t recv_size_ = recv_buf_pair.second;
    
    if (src_size > recv_size_) {
        std::cout << "Dst buffer is too small!" << std::endl;
        exit(-1);
    }

    cudaMemcpyPeerAsync(recv_buf_, peer, src, tid_, src_size, stream);
}

void Worker::recv(int peer, void* recv_buf, size_t recv_size) {
    auto peer_worker = Controller::GetWorker(peer);

    {
        std::lock_guard<std::mutex> lg(mtx_);
        recv_buf_ptr_ = recv_buf;
        recv_buf_size_ = recv_size;
    }

    cv_.notify_all();

    {
        std::unique_lock<std::mutex> ul(peer_worker->mtx___);
        peer_worker->cv___.wait(ul, [peer_worker] { return peer_worker->send_end_;});    
        peer_worker->send_end_ = false;
    }
}

void Worker::send(int peer, const void* src, size_t src_size, cudaStream_t stream) {
    auto peer_worker = Controller::GetWorker(peer);
    void* recv_buf_ptr;
    size_t recv_buf_size;

    {
        std::unique_lock<std::mutex> ul(peer_worker->mtx_);
        peer_worker->cv_.wait(ul, [peer_worker] { return peer_worker->recv_buf_ptr_ != nullptr;});
        recv_buf_ptr = peer_worker->recv_buf_ptr_;
        recv_buf_size = peer_worker->recv_buf_size_;
        
        peer_worker->recv_buf_ptr_ = nullptr;
        peer_worker->recv_buf_size_ = 0;
    }
    
    if (src_size > recv_buf_size) {
        std::cout << "Dst buffer is too small!" << std::endl;
        exit(-1);
    }        

    cudaStreamSynchronize(stream); //ToDo: is this necessary? It seems not to be necessary because cudaMemcpyPeer is a sync-function, but it fails without syncronization.
    cudaMemcpyPeerAsync(recv_buf_ptr, peer, src, tid_, src_size, stream);
    cudaStreamSynchronize(stream);

    {
        std::lock_guard<std::mutex> lg(mtx___);
        send_end_ = true;
    }
    cv___.notify_all();
}

void Worker::send_tensor(torch::Tensor tensor, int dst) {
    auto stream = c10::cuda::getCurrentCUDAStream().stream();
    auto ptr = tensor.data_ptr();
    size_t size = tensor.numel() * torch::elementSize(torch::typeMetaToScalarType(tensor.dtype()));
    send_async(dst, ptr, size, stream);
}

void Worker::recv_tensor(torch::Tensor tensor) {
    auto ptr = tensor.data_ptr();
    size_t size = tensor.numel() * torch::elementSize(torch::typeMetaToScalarType(tensor.dtype()));
    recv_async(ptr, size);
}

void Worker::join() {
    thd_->join();
}

// Class Controller
//public
Controller* Controller::GetInstance() {
    std::call_once(flag_, []() {
        int nr_gpus;
        cudaGetDeviceCount(&nr_gpus);
        instance_ = std::unique_ptr<Controller>(new Controller(nr_gpus));   
        instance_->barrier_init();
        printf("[Manifold] Controller instance created for %d gpus\n", nr_gpus);
});
    
    return instance_.get();
}

Worker* Controller::GetWorker(int tid) {
    return GetInstance()->workers_[tid].get();
}

Worker* Controller::GetCurrentWorker() {
    auto current_ident = std::this_thread::get_id();
    auto ident = std::hash<std::thread::id>{}(current_ident);
    return GetWorkerByIdent(ident);
}

void Controller::join_all() {
    for (auto& worker : workers_) {
        worker->join();
    }
    printf("[Manifold] All workers joined!\n");
}

void Controller::add_idmap(size_t ident, int tid) { // Public for the Worker constructor, but not exported to Python
    idmap_[ident] = tid;
}

void Controller::add_worker(int tid, const std::function<void()>& f) {
    int gpu_id = (nr_gpus_ == 0) ? 0 : tid % nr_gpus_;
    std::cout << "[Manifold] Adding a worker with tid: " << tid << std::endl;
    workers_.emplace_back(std::make_unique<Worker>(tid, gpu_id, f));
}

void Controller::barrier() {
    auto stream = c10::cuda::getCurrentCUDAStream().stream();
    cudaStreamSynchronize(stream);
    pthread_barrier_wait(&barrier_);
}

//private: 
Worker* Controller::GetWorkerByIdent(size_t ident) {
    int tid = GetInstance()->idmap_[ident];
    return GetWorker(tid);
}

void Controller::barrier_init() {
    pthread_barrier_init(&barrier_, NULL, nr_gpus_);
}

} // namespace manifold

File: tensorrt_llm/worker.py
import torch
import threading
from collections import OrderedDict
from cuda import cudart
import queue

def CUDACHECK(cmd):
    err = cmd
    if err[0] != cudart.cudaError_t.cudaSuccess:
        raise RuntimeError("CUDA command returned error: {}".format(err))
    

class Worker:
    def __init__(self, tid, gpu_id, f):
        self.__tid = tid
        self.__gpu_id = gpu_id
        self.recv_queue = queue.Queue()
        self.cv = threading.Condition()
        def thf():
            ident = threading.get_native_id()
            Controller().add_idmap(ident, tid)
            f()
        self.__thd = threading.Thread(target=thf)
        self.__thd.start()
    
    def get_tid(self):
        return self.__tid
    
    def get_gpu_id(self):
        return self.__gpu_id
    
    def join(self):
        self.__thd.join()

    def send_async(self, peer, src, src_size, stream):
        peer_worker = GetWorker(peer)
        recv_buf = 0
        recv_size = 0
        with peer_worker.cv:
            while peer_worker.recv_queue.empty():
                peer_worker.cv.wait()
            recv_buf, recv_size = peer_worker.recv_queue.get()
        
        if src_size > recv_size:
            raise RuntimeError("send_async: src_size > recv_size")
        
        CUDACHECK(cudart.cudaMemcpyPeerAsync(recv_buf, peer, src, self.__gpu_id, src_size, stream.cuda_stream))
        #CUDACHECK(cudart.cudaMemcpyPeer(recv_buf, peer, src, self.__gpu_id, src_size))
    
    def recv_async(self, recv_buf, recv_size):
        with self.cv:
            self.recv_queue.put((recv_buf, recv_size))
            self.cv.notify_all()

    def send_tensor(self, tensor, dst):
        stream = torch.cuda.current_stream()
        tensor_ptr = tensor.data_ptr()
        tensor_size = tensor.numel() * tensor.element_size()
        self.send_async(dst, tensor_ptr, tensor_size, stream)

    def recv_tensor(self, tensor):
        tensor_ptr = tensor.data_ptr()
        tensor_size = tensor.numel() * tensor.element_size()
        self.recv_async(tensor_ptr, tensor_size)
    

class Controller:
    __instance = None
    __instance_lock = threading.Lock()
    __workers = []
    __worker_lock = threading.Lock()
    __idmap = OrderedDict()
    __idmap_lock = threading.Lock()

    def __new__(cls):
        if cls.__instance is None:
            with cls.__instance_lock:
                if cls.__instance is None:
                    cls.__instance = super(Controller, cls).__new__(cls)
                    nr_gpus = torch.cuda.device_count()
                    cls.__instance.__nr_gpus = nr_gpus
                    cls.__instance.__barrier = threading.Barrier(nr_gpus)
                    print("[Manifold] Controller instance created for {} GPUs".format(cls.__instance.__nr_gpus))
        return cls.__instance
    
    def add_worker(self, tid, f):
        self.__worker_lock.acquire()
        gpu_id = 0 if self.__nr_gpus == 0 else tid % self.__nr_gpus
        print("[Manifold] Adding a worker with tid: ", tid)
        self.__workers.append(Worker(tid, gpu_id, f))
        self.__worker_lock.release()

    def add_idmap(self, ident, tid):
        self.__idmap_lock.acquire()
        self.__idmap[ident] = tid
        self.__idmap_lock.release()

    def join_all(self):
        for worker in self.__workers:
            worker.join()
        print("[Manifold] All workers joined")

    def barrier(self, stream_sync=False):
        if stream_sync:
            stream = torch.cuda.current_stream()
            stream.synchronize()
        self.__barrier.wait()

    def get_tid_from_idmap(self, ident):
        self.__idmap_lock.acquire()
        tid = self.__idmap[ident]
        self.__idmap_lock.release()
        return tid

    def get_worker_from_workers(self, tid):
        self.__worker_lock.acquire()
        worker = self.__workers[tid]
        self.__worker_lock.release()
        return worker

def GetWorker(tid):
    return Controller().get_worker_from_workers(tid)

def GetCurrentWorker():
    ident = threading.get_native_id()
    tid = Controller().get_tid_from_idmap(ident)
    return GetWorker(tid)

# For C++
def cuda_send_plugin(peer, src, src_size):
    my_worker = GetCurrentWorker()
    stream = torch.cuda.current_stream()
    stream.synchronize()
    my_worker.send_async(peer, src, src_size, stream)
    Controller().barrier()

def cuda_recv_plugin(recv_buf, recv_size):
    my_worker = GetCurrentWorker()
    my_worker.recv_async(recv_buf, recv_size)
    Controller().barrier()

def barrier_plugin():
    Controller().barrier()